src/python/pants/backend/codegen/protobuf/scala/rules.py

```diff
@@ -0,0 +1,181 @@
+    protoc: Protoc,
+    scalapbc: ScalaPBSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> GeneratedSources:
+    output_dir = "_generated_files"
+
+    downloaded_protoc_binary, tool_classpath, empty_output_dir, transitive_targets = await MultiGet(
+        Get(DownloadedExternalTool, ExternalToolRequest, protoc.get_request(Platform.current)),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(scalapbc.resolved_lockfile(),),
+            ),
+        ),
+        Get(Digest, CreateDigest([Directory(output_dir)])),
+        Get(TransitiveTargets, TransitiveTargetsRequest([request.protocol_target.address])),
+    )
+
+    # NB: By stripping the source roots, we avoid having to set the value `--proto_path`
+    # for Protobuf imports to be discoverable.
+    all_sources_stripped, target_sources_stripped, protoc_gen_scala_digest = await MultiGet(
+        Get(
+            StrippedSourceFiles,
+            SourceFilesRequest(
+                tgt[ProtobufSourceField]
+                for tgt in transitive_targets.closure
+                if tgt.has_field(ProtobufSourceField)
+            ),
+        ),
+        Get(
+            StrippedSourceFiles, SourceFilesRequest([request.protocol_target[ProtobufSourceField]])
+        ),
+        Get(
+            Digest,
+            CreateDigest(
+                [
+                    FileContent(
+                        "protoc-gen-scala",
+                        textwrap.dedent(
+                            f"""\
+                                       #!{bash.path}
+                                       exec {' '.join(jdk_setup.args(bash, tool_classpath.classpath_entries())[1:])} scalapb.ScalaPbCodeGenerator "$@"
+                                       """
+                        ).encode(),
+                        is_executable=True,
+                    )
+                ]
+            ),
+        ),
+    )
+
+    unmerged_digests = [
+        all_sources_stripped.snapshot.digest,
+        downloaded_protoc_binary.digest,
+        tool_classpath.content.digest,
+        protoc_gen_scala_digest,
+        empty_output_dir,
+        jdk_setup.digest,
+    ]
+
+    input_digest = await Get(Digest, MergeDigests(unmerged_digests))
+
+    args = [
+        downloaded_protoc_binary.exe,
+        "--plugin=protoc-gen-scala=./protoc-gen-scala",
+        "--scala_out",
+        output_dir,
+        *target_sources_stripped.snapshot.files,
+    ]
+
+    result = await Get(
+        ProcessResult,
+        Process(
+            args,
+            input_digest=input_digest,
+            description=f"Generating Scala sources from {request.protocol_target.address}.",
+            level=LogLevel.DEBUG,
+            output_directories=(output_dir,),
+            env=jdk_setup.env,
+            append_only_caches=jdk_setup.append_only_caches,
+        ),
```

Comment:
It's probably worth using `nailgun` here, but fine to leave a TODO since this will collide with #13848 anyway.

Reply:
I agree. The main issue is whether we would want to use the `scalapbc` GraalVM-built native binary and let it drive `protoc` invocation and talking with JVM-based protoc plugins. That would likely preclude using nailgun.

Reply:
But I'll get into that this afternoon for the gRPC plugin.

Reply:
> > Added dependency injection based on a `--scalapb-runtime-dependencies` option, but it still requires the user to supply a `jvm_artifact` target to use. Maybe this should support Maven coordinates as well?
> 
> Generally the runtime target should be declared anyway, since it is also used by calling code?

Yes, the user will need to declare that target in any event, so will just update the help message to document that.